tests/test_ensemble.py

```diff
@@ -0,0 +1,203 @@
+        with self.assertRaises(TypeError):
+            EnsembleModel(m)
+        with self.assertRaises(TypeError):
+            EnsembleModel([m, 1])
+        with self.assertRaises(TypeError):
+            EnsembleModel(77)
+        with self.assertRaises(TypeError):
+            EnsembleModel([m, m, m, 77])
+
+    def test_two_models_identical(self):
+        """
+        this tests that the ensemble model works with two identical model-types with identical states, inputs, etc., one with slightly altered parameters
+        """
+        m = OneInputOneOutputOneEventLM()
+        m2 = OneInputOneOutputOneEventLM(x0={'x1': 2})
+
+        # Make sure they're not the same - 3x the impact
+        m2.B = np.array([[3]])
+        m2.C = np.array([[3]])
+        m2.F = np.array([[-0.3]])
+
+        # An ensemble model with two models should work
+        em = EnsembleModel([m, m2])
+
+        # Since they're identical, the inputs, state, etc. should be the same
+        self.assertEqual(em.inputs, m.inputs)
+        self.assertEqual(em.states, m.states)
+        self.assertEqual(em.outputs, m.outputs)
+        self.assertEqual(em.events, m.events)
+
+        # The resulting initial state should be exactly between the two:
+        x0 = em.initialize()
+        self.assertEqual(x0['x1'], 1)
+
+        # Same with state transition
+        u = em.InputContainer({'u1': 1})
+        x1 = em.next_state(x0, u, 1)
+        # m would give 2, m2 would give 4
+        self.assertEqual(x1['x1'], 3)
+
+        # Same with output
+        z = em.output(x1)
+        # m would give 3, m2 would give 9
+        self.assertEqual(z['z1'], 6)
+
+        # Same with event state
+        es = em.event_state(x1)
+        # m would give 0.7, m2 would give 0.1
+        self.assertEqual(es['x1 == 10'], 0.4)
+
+        # performance metrics
+        pm = em.performance_metrics(x1)
+        self.assertEqual(pm['pm1'], 4)
+
+        # Time of event 
+        toe = em.time_of_event(x0, lambda t, x=None: u, dt=1e-3)
+        self.assertAlmostEqual(toe['x1 == 10'], 4.8895)
+
+        # threshold met should be false
+        self.assertFalse(em.threshold_met(x1)['x1 == 10'])
+
+        # Transition again
+        x2 = em.next_state(x1, u, 1)
+        # threshold met should be true (because one of 2 models says it is)
+        self.assertTrue(em.threshold_met(x2)['x1 == 10'])
+
+    def test_two_models_different(self):
+        """
+        this tests that the ensemble model works with two different model-types with different states, inputs, etc. Tests how the ensemble model handles the different values
+        """
+        m = OneInputOneOutputOneEventLM()
+        m2 = OneInputOneOutputOneEventAltLM()
+        em = EnsembleModel([m, m2])
+
+        # inputs, states, outputs, events should be a combination of the two models
+        self.assertSetEqual(set(em.inputs), {'u1', 'u2'})
+        self.assertSetEqual(set(em.states), {'x1', 'x2'})
```

Comment:
Because we use x1 and x2 elsewhere in the code for purposes not only for their states, we should probably either rename our states or rename the later defined variables for x1 and x2.

Reply:
I want to keep the states the same so they're consistent with the other linear test models. Instead, I changed the variable names to be x_t0, x_t1, x_t2, ...